pkg/persistence/config/internal/persistence/type_definition.go

```diff
@@ -71,154 +50,185 @@ type AutomationDefinition struct {
 	if err != nil {
-		return false
+		return fmt.Errorf("failed to unmarshal api-type: %w", err)
 	}
 
-	return a.Name != ""
+	c.Type = config.ClassicApiType{Api: r.Name}
+	c.Scope = r.Scope
+	return nil
 }
-func (c *TypeDefinition) isClassicSound(knownApis map[string]struct{}) error {
-	a, err := UnmarshalApiType(c.Api)
+
+func (c *TypeDefinition) parseSettingsType(a any) error {
+	var r SettingsDefinition
+	err := mapstructure.Decode(a, &r)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to unmarshal settings-type: %w", err)
 	}
 
-	if _, found := knownApis[a.Name]; !found {
-		return errors.New("unknown API: " + a.Name)
+	c.Type = config.SettingsType{
+		SchemaId:      r.Schema,
+		SchemaVersion: r.SchemaVersion,
 	}
+	c.Scope = r.Scope
 	return nil
 }
 
-func (c *TypeDefinition) IsAutomation() bool {
-	return c.Automation != AutomationDefinition{}
+func (c *TypeDefinition) parseAutomation(a any) error {
+	var r AutomationDefinition
+	err := mapstructure.Decode(a, &r)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal automation-type: %w", err)
+	}
+
+	c.Type = config.AutomationType{Resource: r.Resource}
+
+	return nil
 }
 
-func (c *AutomationDefinition) isSound() error {
+// Validate verifies whether the given type definition is valid (correct APIs, fields set, etc)
+func (c *TypeDefinition) Validate(apis map[string]struct{}) error {
+	switch t := c.Type.(type) {
+	case config.ClassicApiType:
+		if _, f := apis[t.Api]; !f {
+			return fmt.Errorf("unknown API: %s", t.Api)
+		}
 
-	switch c.Resource {
-	case "":
-		return errors.New("missing 'type.automation.resource' property")
+	case config.SettingsType:
+		if t.SchemaId == "" {
+			return errors.New("missing settings schemaId")
+		}
 
-	case config.Workflow, config.BusinessCalendar, config.SchedulingRule:
-		return nil
+		if c.Scope == nil {
+			return errors.New("missing settings scope")
+		}
 
-	default:
-		return fmt.Errorf("unknown automation resource %q", c.Resource)
+	case config.AutomationType:
+		switch t.Resource {
+		case "":
+			return errors.New("missing automation resource property")
+
+		case config.Workflow, config.BusinessCalendar, config.SchedulingRule:
+			return nil
+
+		default:
+			return fmt.Errorf("unknown automation resource %q", t.Resource)
+		}
 	}
+
+	return nil
 }
 
-func (c *TypeDefinition) IsBucket() bool {
-	return c.Bucket != ""
+func (c *TypeDefinition) GetApiType() string {
+	switch t := c.Type.(type) {
+	case config.ClassicApiType:
+		return t.Api
+	case config.SettingsType:
+		return t.SchemaId
+	case config.AutomationType:
+		return string(t.Resource)
+	case config.BucketType:
+		return string(t.ID())
+	}
+
+	return ""
 }
 
-func (c *TypeDefinition) GetApiType() (string, error) {
-	switch {
-	case c.IsSettings():
-		return c.Settings.Schema, nil
-	case c.IsClassic():
-		a, err := UnmarshalApiType(c.Api)
-		if err != nil {
-			return "", err
+func (c TypeDefinition) MarshalYAML() (interface{}, error) {
```

Comment:
does this need to be a non-pointer receiver?

Reply:
Yes, it is required